core/src/main/java/com/linecorp/armeria/client/routing/endpoint/healthcheck/HealthCheckedEndpointGroup.java

```diff
@@ -0,0 +1,166 @@
+        this.delegate = requireNonNull(delegate, "delegate");
+        this.healthCheckRetryIntervalMillis = healthCheckRetryIntervalMillis;
+    }
+
+    /**
+     * Update healty servers and start to schedule healthcheck.
+     * A subclass being initialized with this constructor must call {@link #init()} before start being used.
+     */
+    protected void init() {
+        checkAndUpdateHealthyServers().join();
+        scheduleCheckAndUpdateHealthyServers();
+    }
+
+    protected final ClientFactory clientFactory() {
+        return clientFactory;
+    }
+
+    private void scheduleCheckAndUpdateHealthyServers() {
+        clientFactory.eventLoopGroup().schedule(() -> {
+            checkAndUpdateHealthyServers().thenRun(this::scheduleCheckAndUpdateHealthyServers);
+        }, healthCheckRetryIntervalMillis, TimeUnit.MILLISECONDS);
+    }
+
+    private CompletableFuture<Void> checkAndUpdateHealthyServers() {
+        List<ServerConnection> checkedServers = updateServerList();
+
+        CompletableFuture<List<Boolean>> healthCheckResults = CompletableFutures.successfulAsList(
+                checkedServers.stream()
+                              .map(connection -> connection.healthChecker.isHealthy(connection.endpoint()))
+                              .collect(GuavaCollectors.toImmutableList()),
+                t -> false);
+        return healthCheckResults.handle(voidFunction((result, thrown) -> {
+            healthyServers = IntStream
+                    .range(0, result.size())
+                    .filter(i -> result.get(i) != null && result.get(i))
+                    .mapToObj(checkedServers::get)
+                    .collect(GuavaCollectors.toImmutableList());
+        }));
+    }
+
+    /**
+     * Update the servers this health checker client talks to.
+     */
+    private List<ServerConnection> updateServerList() {
+        Map<Endpoint, ServerConnection> allServersByEndpoint = allServers
+                .stream()
+                .collect(GuavaCollectors.toImmutableMap(ServerConnection::endpoint,
+                                                        Function.identity()));
+        return allServers = delegate
+                .endpoints()
+                .stream()
+                .map(endpoint -> {
+                    ServerConnection connection = allServersByEndpoint.get(endpoint);
+                    if (connection != null) {
+                        return connection;
+                    }
+                    return new ServerConnection(endpoint, createEndpointHealthChecker(endpoint));
+                })
+                .collect(GuavaCollectors.toImmutableList());
+    }
+
+    protected abstract EndpointHealthChecker createEndpointHealthChecker(Endpoint endpoint);
+
+    @Override
+    public List<Endpoint> endpoints() {
+        return healthyServers.stream()
+                             .map(ServerConnection::endpoint)
+                             .collect(GuavaCollectors.toImmutableList());
```

Comment:
I guess this method (`endpoints()`) will be invoked pretty often. How about pre-creating it when updating `healthyServers`?